pyproject.toml

```
@@ -0,0 +1,62 @@
+# Prepare a release:
+#
+#  - git pull --rebase
+#  - update version in pyperf/__init__.py and doc/conf.py
+#  - set release date in doc/changelog.rst
+#  - git commit -a -m "prepare release x.y"
+#  - Remove untracked files/dirs: git clean -fdx
+#  - run tests: tox --parallel auto
+#  - git push or send the PR to the repository
+#  - check Github Action CI: https://github.com/psf/pyperf/actions/workflows/build.yml
+#
+# Release a new version:
+#
+#  - go to the GitHub release tab: https://github.com/psf/pyperf/releases
+#  - click "Draft a new release" and fill the contents
+#  - finally click the "Publish release" button! Done!
+#  - monitor the publish status: https://github.com/psf/pyperf/actions/workflows/publish.yml
+#
+# After the release:
+#
+#  - set version to n+1
+#  - git commit -a -m "post-release"
+#  - git push or send the PR to the repository
+
+[build-system]
+requires = ["setuptools >= 35.0.2", "setuptools_scm >= 2.0.0, <3", "wheel"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "pyperf"
+dynamic = ["version"]
+license = {text = "MIT license"}
+description = "Python module to run and analyze benchmarks"
+readme = "README.rst"
+urls = {Homepage = "https://github.com/psf/pyperf"}
+authors= [{name="Victor Stinner", email="vstinner@redhat.com"}]
+maintainers = [{name = "Dong-hee Na", email="donghee.na@python.org"}]
+classifiers = [
+    "Development Status :: 5 - Production/Stable",
+    "Intended Audience :: Developers",
+    "License :: OSI Approved :: MIT License",
+    "Natural Language :: English",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python :: 3",
+    "Topic :: Software Development :: Libraries :: Python Modules"
+]
+dependencies = ["psutil>=5.9.0"]
+
+[project.entry-points.console_scripts]
+pyperf = "pyperf.__main__:main"
+
+[tool.setuptools.dynamic]
+version = {attr = "pyperf.__version__"}
+
+[tool.setuptools]
```

```suggestion
[tool.setuptools]
```

If you specify the version statically above, this section can be eliminated. (It also avoids these sections being out of order.)

---

This was intended, I would like to reduce the defining multiple version semantics.
Currently, we have 3 sections about versioning.
- https://github.com/psf/pyperf/blob/d373c5e56c0257d2d7abd705b676bea25cf66566/setup.py#L27
- https://github.com/psf/pyperf/blob/d373c5e56c0257d2d7abd705b676bea25cf66566/pyperf/__init__.py#L3-L4
- https://github.com/psf/pyperf/blob/d373c5e56c0257d2d7abd705b676bea25cf66566/doc/conf.py#L58

Do you have better ideas?

---

I understand; I just wanted to mention that there is some amount of cost to it. But there's nothing _wrong_ per-say about the current approach; if you're sure that's what you want, what you have would be the way to do it using base Setuptools.

I do suggest, though, moving the `tool.setuptools.dynamic` section below the `tool.setuptools` section to follow standard TOML conventions.

To avoid the duplication in the `conf.py`, since its a dynamic Python file, you can read the version from the `__init__.py` via any number of means—unfortunately, as it imports a bunch of stuff, best not to use `exec` or `importlib`, but you could extract it with `re`, `ast` or other methods.

A popular option for further single-sourcing both the version and other things is Setuptools_scm; the amount of single-sourcing is higher, but also the potential cost and complexities for some types of packaging scenarios. So, likely worth considering separately from this PR.